File: sample_clean_code.py
# Clean Python code example for testing code review tools
# This file demonstrates good coding practices

from typing import List, Optional


def calculate_sum(numbers: List[int]) -> int:
    """Calculate the sum of a list of numbers."""
    return sum(numbers)


def find_max(numbers: List[int]) -> Optional[int]:
    """Find the maximum number in a list."""
    if not numbers:
        return None
    return max(numbers)


class UserRepository:
    """Repository for managing user data."""

    def __init__(self):
        """Initialize the user repository."""
        self._users = []

    def add_user(self, name: str, email: str) -> bool:
        """Add a new user to the repository."""
        if not name or not email:
            return False

        if self._user_exists(email):
            return False

        user = {"name": name, "email": email}
        self._users.append(user)
        return True

    def get_user_by_email(self, email: str) -> Optional[dict]:
        """Get a user by their email address."""
        for user in self._users:
            if user["email"] == email:
                return user
        return None

    def _user_exists(self, email: str) -> bool:
        """Check if a user with the given email already exists."""
        return self.get_user_by_email(email) is not None


def main():
    """Main function to demonstrate the code."""
    numbers = [1, 2, 3, 4, 5]
    print(f"Sum: {calculate_sum(numbers)}")
    print(f"Max: {find_max(numbers)}")

    repo = UserRepository()
    repo.add_user("Alice", "alice@example.com")
    user = repo.get_user_by_email("alice@example.com")
    print(f"User found: {user}")


if __name__ == "__main__":
    main()

File: sample_clean_code.js
// Clean JavaScript code example for testing code review tools
// This file demonstrates good coding practices

/**
 * Calculate the sum of an array of numbers
 * @param {number[]} numbers - Array of numbers to sum
 * @returns {number} The sum of all numbers
 */
function calculateSum(numbers) {
    if (!Array.isArray(numbers)) {
        throw new Error('Input must be an array');
    }
    return numbers.reduce((sum, num) => sum + num, 0);
}

/**
 * Find the maximum number in an array
 * @param {number[]} numbers - Array of numbers
 * @returns {number|null} The maximum number or null if array is empty
 */
function findMax(numbers) {
    if (!Array.isArray(numbers) || numbers.length === 0) {
        return null;
    }
    return Math.max(...numbers);
}

/**
 * Repository class for managing user data
 */
class UserRepository {
    /**
     * Initialize the user repository
     */
    constructor() {
        this._users = [];
    }

    /**
     * Add a new user to the repository
     * @param {string} name - User's name
     * @param {string} email - User's email address
     * @returns {boolean} True if user was added successfully
     */
    addUser(name, email) {
        if (!name || !email) {
            return false;
        }

        if (this._userExists(email)) {
            return false;
        }

        const user = { name, email };
        this._users.push(user);
        return true;
    }

    /**
     * Get a user by their email address
     * @param {string} email - Email address to search for
     * @returns {Object|null} User object or null if not found
     */
    getUserByEmail(email) {
        return this._users.find(user => user.email === email) || null;
    }

    /**
     * Check if a user with the given email already exists
     * @param {string} email - Email address to check
     * @returns {boolean} True if user exists
     * @private
     */
    _userExists(email) {
        return this.getUserByEmail(email) !== null;
    }

    /**
     * Get all users in the repository
     * @returns {Object[]} Array of all users
     */
    getAllUsers() {
        return [...this._users]; // Return a copy to prevent external modification
    }
}

/**
 * Create a user factory function
 * @param {string} type - Type of user to create
 * @returns {Object} User configuration object
 */
function createUser(type) {
    const userConfigs = {
        admin: { type: 'admin', permissions: ['read', 'write', 'delete'] },
        user: { type: 'user', permissions: ['read'] },
        guest: { type: 'guest', permissions: [] }
    };

    return userConfigs[type] || userConfigs.guest;
}

/**
 * Main function to demonstrate the code
 */
function main() {
    const numbers = [1, 2, 3, 4, 5];
    console.log(`Sum: ${calculateSum(numbers)}`);
    console.log(`Max: ${findMax(numbers)}`);

    const repo = new UserRepository();
    repo.addUser('Alice', 'alice@example.com');
    repo.addUser('Bob', 'bob@example.com');

    const user = repo.getUserByEmail('alice@example.com');
    console.log('User found:', user);

    const adminUser = createUser('admin');
    console.log('Admin config:', adminUser);
}

// Run main function if this file is executed directly
if (typeof require !== 'undefined' && require.main === module) {
    main();
}

module.exports = {
    calculateSum,
    findMax,
    UserRepository,
    createUser
};